helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -889,17 +893,23 @@ private int getIdealStateMatched(Partition partition) {
   private void computeIntermediateMap(PartitionStateMap intermediateStateMap,
```

Comment:
To avoid duplicate code, how about generalizing computeIntermediateMap so it takes only one map and always computes with the same logic?

for (Map.Entry<Partition, Map<String, Message>> entry : inputMessageMap.entrySet()) {
      entry.getValue().entrySet().stream().forEach(e -> {
        if (!e.getValue().getToState().equals(HelixDefinedState.DROPPED.name())) {
          intermediateStateMap.setState(entry.getKey(), e.getValue().getTgtName(), e.getValue().getToState());
        } else {
          intermediateStateMap.getStateMap().get(entry.getKey()).remove(e.getValue().getTgtName());
        }
      });
    }

Reply:
@jiajunwang They are different structure of the code. They are different.

Reply:
@jiajunwang They are different structures of the input. One is Map<String, Message> and another is Map<String, String>.